File: leetcode/56.merge-intervals.py
# @leet start
from operator import itemgetter


class Solution:
    def merge(self, intervals: list[list[int]]) -> list[list[int]]:
        intervals.sort(key=itemgetter(0))

        res: list[list[int]] = []
        curr = intervals[0]

        for i in range(1, len(intervals)):
            if intervals[i][0] <= curr[1]:
                curr[1] = max(curr[1], intervals[i][1])
            else:
                res.append(curr)
                curr = intervals[i]

        res.append(curr)

        return res


# @leet end

File: leetcode/56.merge-intervals.js
// @leet start
/**
 * @param {number[][]} intervals
 * @return {number[][]}
 */
var merge = function (intervals) {
  intervals.sort((y1, y2) => {
    if (y1[0] !== y2[0]) {
      return y1[0] - y2[0];
    }

    return y1[1] - y2[1];
  });

  const result = [];
  let currStart = intervals[0][0];
  let currEnd = intervals[0][1];

  for (const [start, end] of intervals) {
    if (start <= currEnd) {
      currEnd = Math.max(currEnd, end);
    } else {
      result.push([currStart, currEnd]);
      currStart = start;
      currEnd = end;
    }
  }

  result.push([currStart, currEnd]);

  return result;
};
// @leet end
